### DISCUSSION TITLE
Ignoring mapping source without using attribute

### DISCUSSION THREAD

**Discussion by joaooo** [MapsterMapper/Mapster#651] [Q&A]
Hello, i'm currently migrating from AutoMapper to Mapster but i'm having some issues with a few things and i was wondering if there wasn't a different way to solve this.

Consider this following example:

```
public class Product 
{
	public string Name { get; set; }
	public Model Model { get; set; }
	public int ModelId { get; set; }
}

public class Model 
{
	public int Id { get; set; }
	public string Name { get; set; }
}

public class ProductDTO 
{
	public string Name { get; set; }
	public Enums.Model Model { get; set; }
}

public enum Model 
{
	Iphone = 1,
	Samsung = 2
}
```

I want to map my ProductDTO.Model from Product.ModelId, in AutoMapper i used to do this:
`.Map(d => d.Model, source => Enum.GetName(typeof(Enums.Model), source.ModelId))`

but when i tried this in Mapster, i get the error "System.InvalidOperationException: Cannot convert immutable type, please consider using 'MapWith' method to create mapping", so i tried to change my code to use .MapWith and returned my DTO but still had the same error, then i realized the error was coming from trying to map my DB Model to the Enum itself and i added this attribute to my Model class: [AdaptIgnore(MemberSide.Source)] and it fixed the error.

I was wondering if there is a different way to accomplish this without having to change my DB Models since i have quite a few of them. I tried to look for workarrounds with .Ignore but i could only ignore the destination class and not the origin.

Thanks in advance!

**Comment by DocSvartz** [comment]
Hello @joaooo  Do you want to map Enum to Class ? 
enum Model  -> class Model

**Reply by DocSvartz** [reply]
```
[TestMethod]
public void EmunDeconstruct()
{
    TypeAdapterConfig<Model651, ModelData651>
        .NewConfig()
        .Map(dest => dest.Id, src => src)
        .Map(dest => dest.Name, src => src);


    var _result = Model651.Samsung.Adapt<ModelData651>();

    _result.Id.ShouldBe(2);
    _result.Name.ShouldBe("Samsung");
    
}

 public class ModelData651
 {
     public int Id { get; set; }
     public string Name { get; set; }
 }

public enum Model651
{
    Iphone = 1,
    Samsung = 2
}

```
Or  This 

```
  TypeAdapterConfig<ProductDTO651, Product651>
      .NewConfig()
      .Map(dest => dest.ModelId, src => src.Model);

  var productDto = new ProductDTO651 { Name = "Galaxy", Model = Model651.Samsung };

  var _productresult = productDto.Adapt<Product651>();

  _productresult.Model.ShouldBe(Model651.Samsung);
  _productresult.Name.ShouldBe("Galaxy");
  _productresult.ModelId.ShouldBe(2); // Model651.Samsung = 2 

```
If You want so that the source does not transmit the value by condition. Then use this
But This case if Type the variable does not match you will get null or default()  in Destination

 `productDto.Adapt<Product651>(); `

```
 TypeAdapterConfig<ProductDTO651, Product651>
     .NewConfig()
     .Map(dest => dest.ModelId, src => src.Model, src=> src.Model.GetType() == typeof(Model651));

```

**Reply by joaooo** [reply]
My apologies, i think i was quite confusing with my post.
My original question was if there wasn't any other way to ignore a property in a source object, besides using 
`[AdaptIgnore(MemberSide.Source)]`
Since i have my Product db model and mapster is trying to match it with my enum, therefore without the attribute, i get an error but it's a pain having to modify quite a lot of auto generated models just to fix this (i also tried to play with .Ignore but could only ignore the destination object and not the source).

Another question if i may, regarding the circular dependencies, does it matter where i place  `.PreserveReference(true)` if it's after the maps or before? Asking this because i'm getting stackoverflow since preserveReference doesn't seem to be working for me

**Reply by DocSvartz** [reply]
.PreserveReference(true) Unfortunately, I can’t tell you about this. But someone already wrote about this error in issue.
You using Mapster.Tool ?

**Reply by DocSvartz** [reply]
https://github.com/MapsterMapper/Mapster/wiki/Rule-based-member-mapping - None of these settings helped?

**Reply by joaooo** [reply]
Regarding the Ignore, none of those really helped, but i ended up implementing with `[AdaptIgnore(MemberSide.Source)]` which fixed the issue.
Regarding the .PreserveReference(true), you mentioned someone had the same issue? Do you have a link? I couldn't really find a lot of issues regarding mapster and circular dependencies. I did not use Mapster.Tool, i wrote my own DTOs. I have tried changing .PreserveReference(true) from the beginning to the end of the config but nothing really helped and i'm still stuck with a stackoverflow exception :/ Not sure if it helps mentioning but i'm trying to migrate from AutoMapper (which works with the circular dependencies)

Also, thanks a lot for all your replies!

**Reply by DocSvartz** [reply]
https://github.com/MapsterMapper/Mapster/issues/571
 .PreserveReference(true)

**Reply by joaooo** [reply]
Thank you! Their example still worked when they removed RequireExplicitMapping, which is odd since i'm not setting mine to true and i'm still getting stackoverflow (sadly i cannot provide my models but it's complex models with a lot of nested objects). I will try once again to play arround .MaxDepth and see if i can make it work, if not, i will just stick with AutoMapper for now.
Thank you so much for your help @DocSvartz :) I appreciate a lot!

**Reply by DocSvartz** [reply]
probably this 
https://github.com/MapsterMapper/Mapster/wiki/Object-references

TypeAdapterConfig.GlobalSettings.Default
 .AvoidInlineMapping(true);
